### PR TITLE
Consolidate docker availability build logic

### PR DESCRIPTION
**Pull request by mark-vieira**
This pull request consolidates and standardizes the mechanisms by which we determine the availability of a compatible Docker installation on the build host, as well as conditionally avoid certain tasks based on that availability. The latter has been refactored such that it is always deferred to task execution or graph calculation time so that in the majority case, when a build does not request any docker-related tasks, we don't make any such attempt to determine if Docker exists on the system.

Prior to this refactoring we have 3 main sources of docker availability logic:

1. The logic in `DockerUtils` which is quite robust and actually checks that Docker exists, can be executed with privileged commands and meets a minimum version requirement. This was used _only_ for the purpose of determining whether we could build the Docker distributions.
1. Some logic in `BuildPlugin` for throwing an exception if any task declared it required a Docker installation meeting the requirements defined in (1). Again, only the `:distribution:docker` project leveraged this.
1. Logic in `DistroTestPlugin` to determine whether or not Docker distribution tests could be executed. This used a different, and more simple implementation than `DockerUtils`, only assuming Windows never works, Mac always works, and Linux works, unless it's a variant that has been explicitly blacklisted in CI.
1. The last bit existed in `TestFixturesPlugins` for the purposes of determining whether `docker-compose` exists on the system. This was another relatively simple implementation, simply looking for the executable from a known list of locations.

This pull request solves some of the inconsistencies listed above via:

1. The introduction of a single `DockerSupportService` that is used in all instances where a determination on whether a compatible Docker installation is needed. This includes building our Docker images, testing them and spinning up Docker-based test fixtures.
1. Some changes to `ElasticsearchDistribution` such that we can depend on a "lenient" distribution. If a distribution sets `required = false` and Docker is unavailable we simply skip building it. This allows testing that uses a locally build Docker distribution to be gracefully skipped in the absence of a Docker installation. This was previously done by checking for Docker availability at configuration time and omitting task dependencies (or entire tasks altogether) to avoid trying to build the Docker images.
1. Refactoring of any direct dependencies on Docker build tasks to use the `DistributionDownloadPlugin` and lenient distributions described in (2).

### CODE REVIEW

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportService.java [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+            availability.version == null ? "" : " v" + availability.version,
+            tasks.size() > 1 ? "s" : "",
+            String.join("\n", tasks),
+            availability.lastCommand.exitCode,
+            availability.lastCommand.stderr.trim()
+        );
+        throwDockerRequiredException(message);
+    }
+
+    private boolean isBlacklistedOs() {
```
Comment (pugnascotia) [inline buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportService.java]: We should stay away from terms like "blacklist" and "whitelist".
```suggestion
    private boolean isExcludedOs() {
```

File: distribution/docker/build.gradle [hunk 2]
```diff
@@ -105,10 +106,19 @@ task copyKeystore(type: Sync) {
...
-preProcessFixture {
-  if (TestFixturesPlugin.dockerComposeSupported()) {
-    dependsOn assemble
+elasticsearch_distributions {
+  Flavor.values().each { distroFlavor ->
+    "docker_$distroFlavor" {
+      flavor = distroFlavor
+      type = 'docker'
+      version = VersionProperties.getElasticsearch()
+      required = false // This ensures we skip this testing if Docker is unavailable
```
Comment (pugnascotia) [inline distribution/docker/build.gradle]: Does this **only** skip testing? Docker is required for a successful overall build on supported OSs, no?

**Review by pugnascotia — APPROVED** [review]
Just one thing I'd like to see changed and one question, but LGTM. This approach looks a lot tidier.

**Comment by mark-vieira** [comment]
@pugnascotia could you provide some context on the necessity of `.ci/dockerOnLinuxExclusions`? Are those machines with Docker installed but something is just borked with it? Would using your logic to actually run a command, test the version, etc be enough to rule out "incompatible" agents or do we still need to explicit exclusions?

File: distribution/docker/build.gradle (hunk 2 above)
Comment (mark-vieira) [reply distribution/docker/build.gradle]: This skips building the upstream docker images that are _consumed_ by the tests. I realize now that comment isn't great, I'll improve it.

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportService.java (hunk 1 above)
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportService.java]: Agreed. I've renamed this.

File: distribution/docker/build.gradle (hunk 2 above)
Comment (mark-vieira) [reply distribution/docker/build.gradle]: I updated the comment text to make it clear this simply makes building the images optional. The logic for actually skipping the tests themselves lives in `TestFixturesPlugin`.

**Review by rjernst — COMMENTED** [review]
This looks great. Just a few comments in addition to the terminology change of "required" to "failifUnavailable" we discussed offline.

File: buildSrc/src/main/java/org/elasticsearch/gradle/test/DistroTestPlugin.java [hunk 3]
```diff
@@ -324,8 +322,14 @@ public String toString() {
...
 
-    private static TaskProvider<?> configureDistroTest(Project project, ElasticsearchDistribution distribution) {
+    private static TaskProvider<?> configureDistroTest(
+        Project project,
+        ElasticsearchDistribution distribution,
+        Provider<DockerSupportService> dockerSupport
+    ) {
         return project.getTasks().register(destructiveDistroTestTaskName(distribution), Test.class, t -> {
+            // Disable Docker distribution tests unless a Docker installation is available
+            t.onlyIf(t2 -> distribution.getType() != Type.DOCKER || dockerSupport.get().getDockerAvailability().isAvailable);
```
[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/test/DistroTestPlugin.java]: As discussed, using `requiresDocker` is _strict_. It fails the build if docker is unavailable. We want to simply skip that task in that scenario.

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportPlugin.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import org.gradle.api.Task;
+import org.gradle.api.plugins.ExtraPropertiesExtension;
+import org.gradle.api.provider.Provider;
+
+import java.io.File;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * <p>Plugin providing {@link DockerSupportService} for detecting Docker installations and determining requirements for Docker-based
```
[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportPlugin.java]: Updated.

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportPlugin.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * Elasticsearch build tasks.</p>
+ *
+ * <p>Additionally registers a task graph listener used to assert a compatible Docker installation exists when task requiring Docker are
```
[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportPlugin.java]: Stupid javadocs. Done.

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportPlugin.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                spec -> spec.parameters(
+                    params -> { params.setExclusionsFile(new File(project.getRootDir(), DOCKER_ON_LINUX_EXCLUSIONS_FILE)); }
+                )
+            );
+
+        // Ensure that if any tasks declare they require docker, we assert an available Docker installation exists
+        project.getGradle().getTaskGraph().whenReady(graph -> {
+            List<String> dockerTasks = graph.getAllTasks().stream().filter(task -> {
+                ExtraPropertiesExtension ext = task.getExtensions().getExtraProperties();
+                return ext.has(REQUIRES_DOCKER_ATTRIBUTE) && ext.get(REQUIRES_DOCKER_ATTRIBUTE).equals("true");
```
[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportPlugin.java]: We were indeed setting to a `boolean` but were checking for a `String`. I've fixed this.

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportService.java [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+
+        return this.dockerAvailability;
+    }
+
+    /**
+     * Given a list of tasks that requires Docker, check whether Docker is available, otherwise throw an exception.
+     *
+     * @throws GradleException if Docker is not available. The exception message gives the reason.
+     */
+    void assertDockerIsAvailable(List<String> tasks) {
```
[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportService.java]: Renamed.

File: buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportService.java [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+        return List.of(DOCKER_COMPOSE_BINARIES).stream().filter(path -> new File(path).exists()).findFirst();
+    }
+
+    private void throwDockerRequiredException(final String message) {
+        throwDockerRequiredException(message, null);
+    }
+
+    private void throwDockerRequiredException(final String message, Exception e) {
+        throw new GradleException(
+            message + "\nyou can address this by attending to the reported issue, " + "removing the offending tasks from being executed.",
```
[the remark this answers is not in the document]
Comment (mark-vieira) [reply buildSrc/src/main/java/org/elasticsearch/gradle/docker/DockerSupportService.java]: I've reworded this message.

**Comment by mark-vieira** [comment]
@rjernst I believe I've addressed you feedback.

File: buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java [hunk 9]
```diff
@@ -270,6 +276,11 @@ void finalizeValues() {
                     "platform not allowed for elasticsearch distribution [" + name + "] of type [" + getType() + "]"
                 );
             }
+            if (getType() == Type.DOCKER && bundledJdk.isPresent()) {
+                throw new IllegalArgumentException(
+                    "bundledJdk not allowed for elasticsearch distribution [" + name + "] of type [docker]"
```
Comment (rjernst) [inline buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: bundledJdk *setting*?

Comment (rjernst) [reply buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java]: or property or something like that? as it is written it sounds like docker can't use the bundledJdk, when in face it only uses it.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by pugnascotia** [comment]
> @pugnascotia could you provide some context on the necessity of `.ci/dockerOnLinuxExclusions`? Are those machines with Docker installed but something is just borked with it? Would using your logic to actually run a command, test the version, etc be enough to rule out "incompatible" agents or do we still need to explicit exclusions?

There were / are some CI machines that can't run Docker for better or worse reasons. I settled on assuming Docker is available unless we know otherwise. I didn't want to rely on detection in case, say, a Packer build went wrong and failed to include Docker or broke it.

**Comment by mark-vieira** [comment]
> I didn't want to rely on detection in case, say, a Packer build went wrong and failed to include Docker or broke it.

Right, I guess my thought is that now that we look for Docker, and try and execute it, we could just rely on that logic for the most part. The risk there though is that should all our Packer builds suddenly stop installing Docker, we'd simply start silently skipping all these tests.

**Comment by rjernst** [comment]
I would rather keep the exclusion list until we can expect docker on all CI images at some future time. We should expect docker to work, and error if it doesn't, except in these edge cases.

**Comment by mark-vieira** [comment]
> I would rather keep the exclusion list until we can expect docker on all CI images at some future time. We should expect docker to work, and error if it doesn't, except in these edge cases.

Works for me. It's not particularly costly to maintain the status quo.

**Pull request merged**